rootston/seat.c

```diff
@@ -495,32 +549,44 @@ void roots_seat_focus_view(struct roots_seat *seat, struct roots_view *view) {
 		return;
 	}
 
-	struct roots_view *prev_focus = seat->focus;
-	seat->focus = view;
+	struct roots_seat_view *seat_view = NULL;
+	if (view != NULL) {
+		bool found = false;
+		wl_list_for_each(seat_view, &seat->views, link) {
+			if (seat_view->view == view) {
+				found = true;
+				break;
+			}
+		}
+		if (!found) {
+			seat_view = seat_add_view(seat, view);
+			if (seat_view == NULL) {
+				wlr_log(L_ERROR, "Allocation failed");
+				return;
+			}
+		}
+	}
+
+	seat->has_focus = false;
 
 	// unfocus the old view if it is not focused by some other seat
-	if (prev_focus && !input_view_has_focus(seat->input, prev_focus)) {
+	if (prev_focus != NULL && !input_view_has_focus(seat->input, prev_focus)) {
 		view_activate(prev_focus, false);
 	}
 
-	if (!seat->focus) {
+	if (view == NULL) {
 		seat->cursor->mode = ROOTS_CURSOR_PASSTHROUGH;
 		return;
 	}
 
-	size_t index = 0;
-	for (size_t i = 0; i < desktop->views->length; ++i) {
-		struct roots_view *_view = desktop->views->items[i];
-		if (_view == view) {
-			index = i;
-			break;
-		}
-	}
-
 	view_activate(view, true);
-	// TODO: list_swap
-	wlr_list_del(desktop->views, index);
-	wlr_list_add(desktop->views, view);
+	wl_list_remove(&seat_view->view->link);
+	wl_list_insert(&seat->input->server->desktop->views,
+		&seat_view->view->link);
```

Comment:
I think it should be raised on the desktop regardless of whether or not it is currently the focused view.